### DISCUSSION TITLE
PCAN-USB X6 not resetting using bus.reset()

### DISCUSSION THREAD

**Discussion by BartoszFEV** [hardbyte/python-can#1668] [Q&A] [answered]
Hello,

I am using a Peak-USB X6 with D-Sub connectors. I am trying to reset a specific channel if `message.is_error_frame`, `message.is_extended_id`, or `message.is_remote_frame` comes back as true. I am currently using `Bus.reset()` to accomplish this. When testing the code on my Vector 1630, I was able to successfully reset only channel 2 after the CAN line was disconnected, and once the communication was re-established the bus became active again and continued on with whatever it was previously doing. Using the exact same code with my X6 however, the bus does not reset. I looked through all the documentation and to my understanding the `PcanBus` object has the function `reset()`. Please take note that I am using this in a class. below is the section of code in question.

```
def run(self):
  print("...Starting to Monitor CAN...")
  while self._ISRUNNING == True:
  try:
    # Read message from bus
    msg = self.canBus.recv(timeout=1.0)
    
    # Catch abnormal messages
    if msg is None or msg.is_error_frame or msg.is_extended_id or msg.is_remote_frame:
      self._BUSEMPTY = True
    
      self.canBus.reset()
    
      print("...Bus Empty Detected!")
      sleep(5)
      continue
  
  except Exception as e: 
    continue
```
                
Can anyone tell me if I am doing something wrong or how I can make this work for my Peak-USB X6 dongle same as it works for my Vector 1630?

**Comment by zariiii9003** [comment, marked as answer]
Looks like `PcanBus.reset` is not correctly implemented. It calls the "CAN_Reset" function of the PCAN Basic API, but that one says `Calling this method ONLY clear the queues of a Channel. A reset of the CAN controller doesn't take place. `

 `PcanBus.reset` should instead uninitialize and initialize the channel.

**Reply by BartoszFEV** [reply]
Are you saying this is an issue with the PCAN Basic API, and that we should shutdown and recreate the bus ourselves to achieve the desired result? If so, would you know the easiest way to implement that?

**Reply by zariiii9003** [reply]
I can't test this, but i tried to fix it here: https://github.com/zariiii9003/python-can/tree/pcanbus_reset

**Reply by BartoszFEV** [reply]
I will test this newer version out as soon as I can and let you know if it worked! Thank you so much sir!

**Reply by BartoszFEV** [reply]
Hello @zariiii9003, I have tested this newer version and it seems to be working as originally intended now. Thank you so much for fixing this issue, if anything changes in the future I will make sure to post on this thread!
